### ISSUE TITLE
aregister miscalculates initial balance (master)

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1634]
In current master (hledger 1.22-103-gc537f9426), `areg -b` is ignoring the begin date and doubling the initial running balance:

    2021-07-26
      (assets)  1000

    $ hledger areg -w80 assets -b 2021-07-31 -e 2021-08-05
    Transactions in assets and subaccounts:
    2021-07-26                      assets                        1000          2000

**Comment by Xitian9** [comment]
As expected, the error is introduced in 6528f2559. I'll look into this.

aregister has very few tests in the suite, so this is probably a good opportunity to introduce more.

**Comment by Xitian9** [comment]
Okay, here's the nub of the problem.

In order to have a proper running balance, `aregister` needs to hold off on filtering transactions until after it has calculated the running balance. But in order for `cur:` and `amt:` to match the pre-valuation currency and amount, rather than the post-valuation currency and amount, we need to filter before doing valuation. I guess a solution is to calculate the running balance with un-valued postings, then value it after all is said and done, but this does not work with `--value=then` valuation, which needs to value before adding all the numbers together.

This is a bit knotty. I think one problem is that I don't have a clear conception of what the running balance should include. It seems that it needs to include all changes to that account, even those that are not displayed in the report. Is this correct?

**Comment by simonmichael** [comment]
IIRC, aregister's running balance should match what you'd see in the bank's online register. This would seem to permit very limited filtering, if any. And yet, https://hledger.org/dev/hledger.html#aregister says "Any additional arguments form a query which will filter the transactions shown.". The example shows only a `date:` query, which is non-problematic. I'll have to play around a bit more to find out.

**Comment by simonmichael** [comment]
If we can "just" get this second test passing again as in 1.22, that would be a good step:
```
<
2021-01-01
  (a)  1

2021-01-02
  (a:aa)  10

2021-01-03
  (a:aa:aaa)  100

# aregister always includes transactions in subaccounts, ignoring a depth limit #1448
$ hledger -f- areg -w80 a -1
Transactions in a and subaccounts:
2021-01-01                      a                                1             1
2021-01-02                      a:aa                            10            11
2021-01-03                      a:aa:aaa                       100           111

# aregister is always in historical mode, showing balance from prior transactions
$ hledger -f- areg -w80 a -b 2021-01-02
Transactions in a and subaccounts:
2021-01-02                      a:aa                            10            11
2021-01-03                      a:aa:aaa                       100           111
```

**Comment by simonmichael** [comment]
And if you prefer me to look at it, let me know (I'm the one who gave it only one test..)

**Comment by simonmichael** [comment]
I guess it allows all query arguments, and then it won't necessarily show the real world running balance, just like `register`.
```
# Any additional arguments are a query, filtering the results. These may cause
# the running balance to diverge from the real-world running balance.
$ hledger -f- areg -w80 a aaa
Transactions in a and subaccounts:
2021-01-03                      a:aa:aaa                       100           100
```

**Comment by Xitian9** [comment]
Here's the situation as of hledger-1.21. Consider both `thisacctq` and `reportq`.
1. Transactions were filtered by `cur:` queries first
2. Next, they were filtered based on any `--real` or `--cleared`/`--pending`/... flags
3. Next, they were filtered based on whether they matche `thisacctq`
4. At this point, they are valued/converted to cost
5. To calculate the historical balance, postings from the above matching `thisacctq`, any non-date parts of `reportq`, and matching a historical query, are included
5. If `--txn-dates` is set, the transactions are then filtered by any other queries
6. At this point, the running balance is calculated based on the historical postings and the transactions still in the queue.
7. To calculate the running balance, filter remaining postings by `reportq`, and add up the amounts of any of those which also match `thisacctq`.

**Comment by simonmichael** [comment]
Fixed by #1636.
